### PR TITLE
Make component examples an optional Elyra feature

### PR DESCRIPTION
**Pull request by ptitzler**
This PR
 - ~requires merge of https://github.com/elyra-ai/examples/pull/79~ **RESOLVED**
 - ~requires merge of https://github.com/elyra-ai/elyra/pull/2282~ **RESOLVED**
 - ~requires merge of https://github.com/elyra-ai/elyra/pull/2263~ **RESOLVED**
 - closes https://github.com/elyra-ai/elyra/pull/2272
 - closes https://github.com/elyra-ai/elyra/issues/2102
 - closes https://github.com/elyra-ai/elyra/issues/2164

**New user experience**

- after running `pip install elyra` (`make install`) in a clean environment
   - the component catalog list is empty
    - the KFP VPE palette contains  only generic components (in the `Elyra` category) 
    - the AA VPE palette contains  generic components (in the `Elyra` category) 
   - the user can manually install the example components at a later time, as outlined in  the documentation provided in https://github.com/elyra-ai/examples/pull/79
  
- after running `pip install elyra[all]` (`make install-all`) in a clean environment
    - the component catalog list is empty
    - the KFP VPE palette contains  only generic components (in the `Elyra` category) 
    - the AA VPE palette contains  generic components (in the `Elyra` category) 
    - the user can enable the Kubeflow Pipelines examples by creating a new "Kubeflow Pipelines example catalog" (via the GUI or CLI)
    - the user can enable the Apache Airflow examples by creating a new "Apache Airflow example catalog" (via the GUI or CLI)
    - the examples can be removed from the catalog list (and thus the VPE palette) or re-added after removal

- after launching the pre-built `elyra` container image
    - the component catalog lists two entries - one for KFP examples and one for AA examples
    - the KFP VPE palette contains  generic components (in the `Elyra` category) and the KFP examples (in the `examples` category)
    - the AA VPE palette contains  generic components (in the `Elyra` category) and the AA examples (in the `examples` category)
    - users can customize the container image `Dockerfile` as desired (with respect to making examples available or not)

- after launching the pre-built `kf-notebook` container image
    - the component catalog lists one entry for KFP examples 
    - the KFP VPE palette contains  generic components (in the `Elyra` category) and the KFP examples (in the `examples` category)
    - the examples can be removed from the catalog list (and thus the VPE palette) or re-added after removal
    - users can customize the container image `Dockerfile` as desired (with respect to making examples available or not)

### What changes were proposed in this pull request?

This PR
 - removes the Kubeflow Pipelines and Apache Airflow component examples from the Elyra code base
 - updates the `runtime_extras` in `setup.py` to require the Kubeflow Pipelines and Apache Airflow component catalogs, which make the examples available (again)
 - installs (and enables) the Kubeflow Pipelines components examples in the Kubeflow Dockerfile
 - installs (and enables) the Kubeflow Pipelines and Apache Airflow components examples in the Elyra Dockerfile
 - requires `.pipeline` file migration (details are in https://github.com/elyra-ai/pipeline-editor/pull/172)
 - adds a new `install-all` target to the `Makefile`, which depends on `install` and `install-examples`
 - adds a new `install-examples` target to the `Makefile`, which installs the KFP and AA example component packages from PyPI
 - Updates the 'pipeline components' topic in the 'user guide'

### How was this pull request tested?

 
Developer's Certificate of Origin 1.1

       By making a contribution to this project, I certify that:

       (a) The contribution was created in whole or in part by me and I
           have the right to submit it under the Apache License 2.0; or

       (b) The contribution is based upon previous work that, to the best
           of my knowledge, is covered under an appropriate open source
           license and I have the right under that license to submit that
           work with modifications, whether created in whole or in part
           by me, under the same open source license (unless I am
           permitted to submit under a different license), as indicated
           in the file; or

       (c) The contribution was provided directly to me by some other
           person who certified (a), (b) or (c) and I have not modified
           it.

       (d) I understand and agree that this project and the contribution
           are public and that a record of the contribution (including all
           personal information I submit with it, including my sign-off) is
           maintained indefinitely and may be redistributed consistent with
           this project or the open source license(s) involved.

### CODE REVIEW

**Comment by ptitzler** [comment]
Test failures:
```
FAILED elyra/tests/cli/test_pipeline_app.py::test_validate_with_kfp_components - assert 1 == 0
FAILED elyra/tests/pipeline/test_validation.py::test_invalid_runtime_node_kubeflow - assert 2 == 1
FAILED elyra/tests/pipeline/test_validation.py::test_invalid_runtime_node_kubeflow_with_supernode - assert 3 == 1
FAILED elyra/tests/pipeline/test_validation.py::test_missing_node_property_for_kubeflow_pipeline - KeyError: 'c...
FAILED elyra/tests/pipeline/test_validation.py::test_pipeline_kfp_inputpath_parameter - KeyError: 'current_para...
FAILED elyra/tests/pipeline/test_validation.py::test_pipeline_invalid_kfp_inputpath_parameter - KeyError: 'curr...
FAILED elyra/tests/pipeline/test_validation.py::test_pipeline_invalid_kfp_inputpath_missing_connection - KeyErr...
FAILED elyra/tests/pipeline/airflow/test_component_parser_airflow.py::test_component_registry_can_load_components_from_registries
FAILED elyra/tests/pipeline/kfp/test_component_parser_kfp.py::test_component_registry_can_load_components_from_registries
ERROR elyra/tests/pipeline/airflow/test_processor_airflow.py::test_create_file_custom_components[resources/sample_pipelines/pipeline_with_airflow_components.json]
```

File: etc/docker/elyra/Dockerfile [hunk 1]
```diff
@@ -48,5 +48,11 @@ RUN conda remove --force -y terminado && \
         jupyter lab build; \
     fi
 
+# Comment out to hide the component examples for Kubeflow Pipelines and Apache Airflow
+RUN if [ "$TAG" != "dev" ] ; then \
+        elyra-metadata install component-registries --schema_name=elyra-airflow-examples-catalog --display_name="Apache Airflow examples" --runtime_type=APACHE_AIRFLOW --categories="['examples']" --version=1 && \
```
Comment (ptitzler) [inline etc/docker/elyra/Dockerfile:53]: TODO after https://github.com/elyra-ai/elyra/pull/2282 and https://github.com/elyra-ai/elyra/pull/2263 were merged

```suggestion
        elyra-metadata install component-catalogs --schema_name=elyra-airflow-examples-catalog --display_name="Apache Airflow examples" --runtime_type=APACHE_AIRFLOW --categories="['examples']" && \
```

**Comment by kiersten-stokes** [comment]
A few notes re: the current test failures:

- I think we may want to consider removing the test `test_component_registry_can_load_components_from_registries` for both Airflow and KFP because this functionality is already covered by the following test (`test_modify_component_registries`), and we will no longer have any built-in components
- The remaining 8 failures are due to the lack of example components that are available out of the box now -- we'll have to discuss whether we want to maybe Mock a Component cache in these situations or some other solution 
  - Note that the Component cache structure will be changing in 3.4 (due to #2162), which will affect all of these tests as well. We may want to consider this in deciding on a course of action

File: conftest.py [hunk 2]
```diff
@@ -15,5 +15,48 @@
 #
 import pytest
 
+from elyra.metadata.metadata import Metadata
+from elyra.metadata.schemaspaces import ComponentCatalogs
+
```
Comment (kiersten-stokes) [inline conftest.py:20]: Remove unnecessary line

File: Makefile [hunk 3]
```diff
@@ -140,6 +144,17 @@ install-server-package:
 
 install: install-server install-ui check-install ## Build and install
 
+install-all: install-server install-ui check-install install-examples ## Build and install, including examples
```
Comment (akchinSTC) [inline Makefile:147]: Seeing weird behavior here, running `make install` installs the examples packages. Is it supposed to do this? I figure the `install` target should be installing elyra but not the extra packages, hence the two separate targets here for all and not all.

```
(python38) Alans-MacBook-Pro:elyra akchin$ pip list installed | grep elyra
elyra                             3.3.0.dev0
elyra-examples-airflow-catalog    0.0.1
elyra-examples-kfp-catalog        0.0.1
```

**Review by kiersten-stokes — APPROVED** [review]

**Comment by ptitzler** [comment]
> Seeing weird behavior here, running make install installs the examples packages.

It appears the examples are installed through `test_requirements.txt`

```
$ make install-server --just-print
python -m pip install --upgrade pip
pip install -q -r test_requirements.txt
flake8 elyra
pip install --upgrade pip wheel
python setup.py bdist_wheel sdist
pip install --upgrade --upgrade-strategy only-if-needed --use-deprecated=legacy-resolver "dist/elyra-3.3.0.dev0-py3-none-any.whl[kfp-tekton]"
```

```
$ cat test_requirements.txt 
elyra-examples-airflow-catalog
elyra-examples-kfp-catalog
...
```

**Comment by ptitzler** [comment]
Waiting for tests to complete. Do note that once somebody runs implicitly or explicitly `make test-server` the environment is 'contaminated' and all sample components are installed irrespective of whether one ran `make install` or `make install-all`.

**Review by akchinSTC — APPROVED** [review]

**Comment by ptitzler** [comment]
Confirmed using the 3.0.0rc0 `elyra` package that `/etc/docker/elyra/Dockerfile` enables the example components for KFP and AA as expected

**Comment by ptitzler** [comment]
Confirmed using the 3.0.0rc0 `elyra` package that `/elyra/etc/docker/kubeflow/Dockerfile` enables the example components for KFP as expected

**Pull request merged**